Caption relative timestamps rebase

JacksonRG opened the pull request:
# Issue:
Timestamps took absolute times for when to start and stop displaying.

This meant that they had to be adjusted any time a clip was moved.

# Solution:
Subtract the `position` (the time at the left edge of a clip) before creating the timestamp.

Note: `libopenshot`'s caption code already accounted for the beginning clips, so no change to libopenshot is needed.

Other Improvements:
- Prevent timestamps outside the time of the clip
- Prevent `add timestamp` button from adding after a line has an `ending` timestamp

Commit:
Make caption timestamps relative to the start of the clip

Rebasing these changes onto develop.

Commit:
Multiple updates to caption effect

- Prevent captions before or after the clip
- Don't add repeatedly add `ending` timestamps
- Make timestamps represent a position in time within the clip

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
         _ = app._tr
 
+        if len(self.selected_effects) == 0:
+            log.error("No caption effect selected")
```
I think this might send data to Sentry.io, probably 'error' is to strong

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
+        effect_data = Effect.filter(id=self.selected_effects[0])[0].data
+        effect_id = effect_data.get("id")
+        if effect_data.get("type") != "Caption":
+            log.error("Captioning an effect that is not a Caption")
```
ditto

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
+                break
+
+        if clip_data == None:
+            log.error("No clip owns this caption effect")
```
Ditto

Commit:
Use log to info not error for problems finding the clip a caption belongs to.

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
         app = get_app()
         _ = app._tr
 
+        if len(self.selected_effects) == 0:
```
`if self.selected_effects:` does the same thing

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
+
+        # Get the Clip that owns this caption effect
+        clip_data = None
+        for clip in Clip.filter():
```
Clip.filter() will return all clips in the project. It would probably be better if you could get the Clip ID of the parent clip, and pass it to the filter() function. Not sure if we have the Clip ID in the effect data though.

JacksonRG replied:
I just checked `clip_data.get("id") in str(effect_data)` -> `false`. Not seeing a way to get clip id out of the effect

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
+        if clip_data == None:
+            log.info("No clip owns this caption effect")
+            return
+
```
Extra blank line

jonoomph commented on src/windows/main_window.py:
```diff
@@ -2232,25 +2232,65 @@ def actionInsertTimestamp_trigger(self, event):
+        # If position hasn't changed, and creating the ending timestamp, add 5 seconds
+        if "-->"  in line_text and line_text.count(':') == 3 and self.last_caption_timestamp_position == current_position:
+            # prevent caption with 0 duration
+            relative_position += 5
```
Should this be `+= 5.0`?

jonoomph commented:
I feel like it should be easier to get the Clip ID from the selected effect. But off the top of my head, I don't recall. Otherwise looks good.

Commit:
Suggested edits, and fixes for cut clips.

- Code cleanup
- Caption timestamps take trimming into account.
- Don't jump to the end if cursor is on a blank line.

jonoomph commented:
LGTM
